pyrate/core/dem_error.py

```diff
@@ -0,0 +1,307 @@
+
+    # incidence angle at pixel ij -> law of cosines in "satellite - Earth centre - ground pixel" triangle
+    # see e.g. Section 2 in https://www.cs.uaf.edu/~olawlor/ref/asf/sar_equations_2006_08_17.pdf
+    incidence_angle = np.pi - np.arccos(np.divide(np.square(range) + np.square(re) - se**2, 2 * np.multiply(range, re)))
+    # for validation with GAMMA output only
+    incidence_angle_gamma = np.pi / 2 - incidence_angle
+    # maximum differences to the GAMMA-derived local incidence angles for Sentinel-1 test data are within +/-0.1 deg
+
+    # local azimuth angle at pixel ij using constant satellite heading angle and spherical approximations
+    epsilon = np.pi - look_angle - (np.pi - incidence_angle) # angle at the Earth's center between se and re
+    # azimuth of satellite look vector(for right-looking SAR)
+    sat_azi = heading + np.pi / 2
+    # the following equations are adapted from Section 4.4 (page 4-16) in EARTH-REFERENCED AIRCRAFT NAVIGATION AND
+    # SURVEILLANCE ANALYSIS (https://ntlrepository.blob.core.windows.net/lib/59000/59300/59358/DOT-VNTSC-FAA-16-12.pdf)
+    sat_lon = np.divide(np.arcsin(-(np.multiply(np.sin(epsilon), np.sin(sat_azi)))), np.cos(lat)) + lon # Eq. 103
+    temp = np.multiply(np.divide(np.cos(0.5 * (sat_azi + sat_lon - lon)), np.cos(0.5 * (sat_azi - sat_lon + lon))), \
+                       np.tan(0.5 * (np.pi / 2 + lat - epsilon))) # Eq. 104
+    sat_lat = -np.pi / 2 + 2 * np.arctan(temp)
+    # the above code could be improved by calculating satellite positions for each azimuth row
+    # from orbital state vectors given in .par file, using the following workflow:
+    # 1. read orbital state vectors and start/stop times from mli.par
+    # 2. for each pixel get the corresponding radar row (from matrix az)
+    # 3. get the corresponding radar time for that row (using linear interpolation)
+    # 4. calculate the satellite XYZ position for that time by interpolating the time and velocity state vectors
+
+    # calc azimuth angle using Vincenty's equations
+    azimuth_angle = np.empty(lat.shape) # pre-allocate 2D numpy array
+    azimuth_angle[:] = np.nan
+    for ix, iy in np.ndindex(lat.shape):
+        if not isnan(sat_lat[ix, iy]):
+            az12 = vincinv(lat[ix, iy], lon[ix, iy], sat_lat[ix, iy], sat_lon[ix, iy], a, b)
+            azimuth_angle[ix, iy] = az12
+    np.reshape(azimuth_angle, lat.shape)
+
+    # old code: spherical azimuth angle calculation
+    #azimuth_angle2 = np.arccos(np.divide(np.multiply(np.sin(sat_lat), np.cos(lat)) - \
+    #                                np.multiply(np.multiply(np.cos(sat_lat), np.sin(lat)), np.cos(sat_lon - lon)), \
+    #                                np.sin(epsilon)))
+    #azimuth_angle_diff = azimuth_angle2 - azimuth_angle
+    #print(np.nanmin(azimuth_angle_diff), np.nanmax(azimuth_angle_diff))
+    # the difference between Vincety's azimuth calculation and the spherical approximation is ~0.001 radians
```

Comment:
delete this if it is redundant. potentially you could use it in a test for validation of the preferred method?

Reply:
Good point. Added a todo comment to move this to tests.